cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,681 @@
+
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        get_dataframe_buffer_begin(vertex_pair_buffer_p_r_edge_p_q),
+        get_dataframe_buffer_end(vertex_pair_buffer_p_r_edge_p_q),
+        generate_p_r_q_r<vertex_t,
+                         edge_t,
+                         decltype(get_dataframe_buffer_begin(invalid_edges_buffer))>{
+          0,
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+          raft::device_span<vertex_t const>(intersection_indices.data(),
+                                            intersection_indices.size()),
+          get_dataframe_buffer_begin(invalid_edges_buffer)});
+
+      auto edge_last       = transposed_edge_first + edgelist_srcs.size();
+      auto num_edge_exists = accumulate_pair_size;
+      thrust::for_each(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator<edge_t>(0),
+                       thrust::make_counting_iterator<edge_t>(num_edge_exists),
+                       unroll_edge<vertex_t, edge_t, decltype(transposed_edge_first)>{
+                         raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+                         get_dataframe_buffer_begin(vertex_pair_buffer_p_r_edge_p_q),
+                         transposed_edge_first,
+                         edge_last,
+                       });
+
+      auto vertex_pair_buffer_q_r_edge_p_q =
+        allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(accumulate_pair_size,
+                                                                     handle.get_stream());
+
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        get_dataframe_buffer_begin(vertex_pair_buffer_q_r_edge_p_q),
+        get_dataframe_buffer_begin(vertex_pair_buffer_q_r_edge_p_q) + accumulate_pair_size,
+        generate_p_r_q_r<vertex_t,
+                         edge_t,
+                         decltype(get_dataframe_buffer_begin(invalid_edges_buffer))>{
+          1,
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+          raft::device_span<vertex_t const>(intersection_indices.data(),
+                                            intersection_indices.size()),
+          get_dataframe_buffer_begin(invalid_edges_buffer)  // FIXME: verify this is accurate
+        });
+
+      thrust::for_each(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator<edge_t>(0),
+                       thrust::make_counting_iterator<edge_t>(num_edge_exists),
+                       unroll_edge<vertex_t, edge_t, decltype(transposed_edge_first)>{
+                         raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+                         get_dataframe_buffer_begin(vertex_pair_buffer_q_r_edge_p_q),
+                         transposed_edge_first,
+                         edge_last,
+                       });
+    }
```

Comment:
You need to reduce (p,q) edges' triangle counts by the size of the intersections as well.